akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -0,0 +1,217 @@
+/**
+ *  Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.dispatch.affinity
+
+import java.util
+import java.util.concurrent._
+import java.util.concurrent.locks.{ Lock, ReentrantLock }
+import akka.dispatch.{ AbstractBoundedNodeQueue, DispatcherPrerequisites, ExecutorServiceConfigurator, ExecutorServiceFactory }
+import akka.dispatch.affinity.AffinityPool._
+import com.typesafe.config.Config
+import net.openhft.affinity.{ AffinityStrategies, AffinityThreadFactory }
+
+import scala.collection.mutable
+import scala.collection.JavaConversions._
+import scala.util.{ Failure, Try }
+
+class AffinityPool(numThreads: Int, affinityGroupSize: Int, tf: ThreadFactory) extends AbstractExecutorService {
+
+  // controls access to mutable such as list of workers, etc
+  private val mainLock = new ReentrantLock
+  @volatile private var poolState: PoolState = Running
+
+  // there is one queue for each thread
+  private val workQueues = Array.fill(numThreads)(new BoundedTaskQueue(affinityGroupSize))
+
+  //mutable set of all workers that are currently running
+  private val workers: mutable.Set[ThreadPoolWorker] = mutable.Set()
+
+  //fires up initial workers
+  workQueues.foreach(q ⇒ addWorker(workers, q))
+
+  private def addWorker(workers: mutable.Set[ThreadPoolWorker], q: BoundedTaskQueue): Unit = {
+    locked(mainLock) {
+      val worker = new ThreadPoolWorker(q)
+      workers.add(worker)
+      worker.startWorker()
+    }
+  }
+
+  private def tryEnqueue(command: Runnable) = {
+    // does a mod on the hash code to determine which queue to put task in
+    val queueIndex = Math.abs(command.hashCode()) & (numThreads - 1)
+    workQueues(queueIndex).add(command)
+  }
+
+  private def onWorkerExit(w: ThreadPoolWorker): Unit = {
+    locked(mainLock) {
+      workers.remove(w)
+    }
+  }
+
+  private def onWorkerFailure(w: ThreadPoolWorker): Unit = {
+    locked(mainLock) {
+      workers.remove(w)
+      addWorker(workers, w.q)
+    }
+  }
+
+  private class ThreadPoolWorker(val q: BoundedTaskQueue) extends Runnable {
+
+    sealed trait WorkerState
+    case object NotStarted extends WorkerState
+    case object InExecution extends WorkerState
+    case object Idle extends WorkerState
+
+    private val thread = tf.newThread(this)
+    @volatile var workerState: WorkerState = NotStarted
+
+    def startWorker() = thread.start()
+
+    def runCommand(command: Runnable) = {
+      workerState = InExecution
+      try
+        command.run()
+      catch {
+        case x: RuntimeException ⇒ throw x
+        case e: Error            ⇒ throw e
+        case t: Throwable        ⇒ throw new Error(t)
+      } finally {
+        workerState = Idle
+      }
+    }
+
+    override def run(): Unit =
+      Try {
```

Comment:
It's recommended use the regular `try` construct if you do not intend of passing the result around.
And you'll want to use `scala.util.NonFatal`.